### PR TITLE
SOLR-17431: Deprecate -p parameter where it doesn't refer to a port.

### PR DESCRIPTION
**Pull request by epugh**
https://issues.apache.org/jira/browse/SOLR-17431

# Description

Deprecate -p for all options where it doesnt' refer to a port.

# Solution

add deprecation warnings.

# Tests

ran tests

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended, not available for branches on forks living under an organisation)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Review by malliaridis — CHANGES_REQUESTED** [review]
A few more places I have spotted `-p`:
- [post-tool.adoc](https://github.com/epugh/solr/blob/SOLR-17431/solr/solr-ref-guide/modules/indexing-guide/pages/post-tool.adoc?plain=1#L77)
- [PackageToolTest.java#L170](https://github.com/epugh/solr/blob/SOLR-17431/solr/core/src/test/org/apache/solr/cli/PackageToolTest.java#L170)
- [PackageToolTest.java#L266](https://github.com/epugh/solr/blob/SOLR-17431/solr/core/src/test/org/apache/solr/cli/PackageToolTest.java#L266)

For backporting to `branch_9X` there is also `PackageManagerCLITest` with a few occurences (as far as I have found).

Besides that, PR is ready. :)

File: solr/core/src/java/org/apache/solr/cli/ConfigTool.java [hunk 1]
```diff
@@ -93,9 +109,13 @@ public void runImpl(CommandLine cli) throws Exception {
     String solrUrl = SolrCLI.normalizeSolrUrl(cli);
     String action = cli.getOptionValue("action", "set-property");
     String collection = cli.getOptionValue("name");
-    String property = cli.getOptionValue("property");
+    String property = SolrCLI.getOptionWithDeprecatedAndDefault(cli, "property", "p", null);
     String value = cli.getOptionValue("value");
 
+    if (property == null) {
+      throw new IllegalArgumentException("'property' is a required option.");
```
[the remark this answers is not in the document]
Comment (epugh) [reply solr/core/src/java/org/apache/solr/cli/ConfigTool.java:116]: why not!

File: solr/core/src/java/org/apache/solr/cli/SolrCLI.java [hunk 2]
```diff
@@ -371,6 +371,18 @@ public static Options getToolOptions(Tool tool) {
     return options;
   }
 
+  /**
+   * Returns the value of the option with the given name, or the value of the deprecated option with
```
Comment (janhoy) [inline solr/core/src/java/org/apache/solr/cli/SolrCLI.java:375]: Incomplete sentence, but that's also true in 9x

**Review by malliaridis — APPROVED** [review]

File: solr/core/src/test/org/apache/solr/cli/PackageToolTest.java [hunk 3]
```diff
@@ -263,7 +264,7 @@ public void testPackageTool() throws Exception {
               "question-answer",
               "--collections",
               "abc",
-              "-p",
+              "--params",
```
Comment (janhoy) [inline solr/core/src/test/org/apache/solr/cli/PackageToolTest.java:267]: Should not this be `--param`?

Comment (epugh) [reply solr/core/src/test/org/apache/solr/cli/PackageToolTest.java:267]: It should, but we're going to do a separate pass on all the variations of this property!  Checking the spreadsheet we have:  
```

-params
--
--params
--param
--parameters
```

Comment (epugh) [reply solr/core/src/test/org/apache/solr/cli/PackageToolTest.java:267]: Package tool has: ```
Option.builder()
            .longOpt("param")
            .hasArgs()
            .argName("PARAMS")
            .desc("List of parameters to be used with deploy command.")
            .build(),
```

**Comment by epugh** [comment]
@janhoy  can I get an LGTM on this?

**Review by janhoy — APPROVED** [review]

**Pull request merged**